crates/aptos-crypto/src/webauthn/webauthn_p256_keys.rs

```diff
@@ -0,0 +1,280 @@
+    traits::*,
+    webauthn::webauthn_p256_sigs::WebAuthnP256Signature,
+};
+use aptos_crypto_derive::{DeserializeKey, SerializeKey, SilentDebug, SilentDisplay};
+use core::convert::TryFrom;
+use p256::{self, ecdsa};
+#[cfg(any(test, feature = "fuzzing"))]
+use proptest::prelude::*;
+use serde::Serialize;
+use std::fmt;
+
+/// A WebAuthn P256 private key
+#[derive(DeserializeKey, SerializeKey, SilentDebug, SilentDisplay)]
+pub struct WebAuthnP256PrivateKey(pub(crate) P256PrivateKey);
+
+impl private::Sealed for WebAuthnP256PrivateKey {}
+
+#[cfg(feature = "assert-private-keys-not-cloneable")]
+static_assertions::assert_not_impl_any!(WebAuthnP256PrivateKey: Clone);
+
+#[cfg(any(test, feature = "cloneable-private-keys"))]
+impl Clone for WebAuthnP256PrivateKey {
+    fn clone(&self) -> Self {
+        let serialized: &[u8] = &(self.to_bytes());
+        WebAuthnP256PrivateKey::try_from(serialized).unwrap()
+    }
+}
+
+/// A WebAuthn P256 public key
+#[derive(DeserializeKey, Clone, SerializeKey)]
+pub struct WebAuthnP256PublicKey(pub(crate) P256PublicKey);
+
+impl private::Sealed for WebAuthnP256PublicKey {}
+
+impl WebAuthnP256PrivateKey {
+    /// The length of the WebAuthnP256PrivateKey
+    pub const LENGTH: usize = P256PrivateKey::LENGTH;
+
+    /// Serialize a WebAuthnP256PrivateKey.
+    pub fn to_bytes(&self) -> [u8; P256PrivateKey::LENGTH] {
+        self.0.to_bytes()
+    }
+
+    /// Deserialize an WebAuthnP256PrivateKey without any validation checks apart from expected key size.
+    fn from_bytes_unchecked(
+        bytes: &[u8],
+    ) -> std::result::Result<WebAuthnP256PrivateKey, CryptoMaterialError> {
+        match p256::ecdsa::SigningKey::from_slice(bytes) {
+            Ok(p256_secret_key) => Ok(WebAuthnP256PrivateKey(P256PrivateKey(p256_secret_key))),
+            Err(_) => Err(CryptoMaterialError::DeserializationError),
+        }
+    }
+
+    /// This may be problematic
+    /// Trait method is required but DOES NOT work well for WebAuthn signatures
+    /// because verification_data is unknown without auth_data and client_data_json
+    ///
+    /// Private function aimed at minimizing code duplication between sign
+    /// methods of the SigningKey implementation. This should remain private.
+    fn sign_arbitrary_message(&self, message: &[u8]) -> WebAuthnP256Signature {
+        let secret_key: &ecdsa::SigningKey = &self.0 .0;
+        WebAuthnP256Signature {
+            actual_challenge: vec![],
+            verification_data: vec![],
+            p256_signature: P256Signature(<ecdsa::SigningKey as signature::Signer<
+                ecdsa::Signature,
+            >>::sign(secret_key, message)),
+        }
+    }
+}
+
+impl WebAuthnP256PublicKey {
+    /// Serialize a WebAuthnP256PublicKey.
+    pub fn to_bytes(&self) -> [u8; P256_PUBLIC_KEY_LENGTH] {
+        // The RustCrypto P256 `to_sec1_bytes` call here should never return an array of the wrong length and cause a panic
```

Comment:
This doesn't appear to use `to_sec1_bytes`

Reply:
Removed comment